### PR TITLE
Extension of correction-lib implementation to Run 3 SFs -- topcoffea

### PR DESCRIPTION
**Pull request by anpicci**
With this PR (already merged with @MatthewDittrich 's PR), I have both concluded the extension to Run3, and harmonized the scripts in order to have a more transparent approach at the `analysis_processor` level.

### CODE REVIEW

File: topcoffea/modules/event_selection.py [hunk 1]
```diff
@@ -9,7 +9,10 @@
 def passes_trg_inlst(events,trg_name_lst):
     tpass = np.zeros_like(np.array(events.MET.pt), dtype=bool)
     trg_info_dict = events.HLT
-
+    #print("\n\n\n\n\n\n")
```
Comment (kmohrman) [inline topcoffea/modules/event_selection.py:12]: Is this needed? It looks like maybe a leftover debugging statement that could be removed.

File: topcoffea/params/xsec.yml [hunk 2]
```diff
@@ -209,4 +209,47 @@ GluGlutoContinto2Zto4TauK_13p6TeV       : 0.0051
 TbarWplustoLNu2Q_13p6TeV           : 19.36
 TWminustoLNu2Q_13p6TeV             : 19.36
 
+#NEW from multileptons for Run3
```
Comment (kmohrman) [inline topcoffea/params/xsec.yml:212]: We will need a comment with references for these (like a link, or explanation pointing to where they came from)

File: topcoffea/modules/createJSON.py [hunk 3]
```diff
@@ -85,7 +85,10 @@ def main():
         dicFiles = GetDatasetFromDAS(dataset, nFiles, options='file', withRedirector=prefix)
         files = [f[len(prefix):] for f in dicFiles['files']]
         files_with_prefix = dicFiles['files']
-
+        if 'root://cms-xrd-global.cern.ch//store/mc/Run3Summer22NanoAODv12/ZZZ_TuneCP5_13p6TeV_amcatnlo-pythia8/NANOAODSIM/130X_mcRun3_2022_realistic_v5-v2/50000/7c4f3eb2-3c7e-4c21-98ed-c1892bb3a057.root'\
```
Comment (kmohrman) [inline topcoffea/modules/createJSON.py:88]: Why is this needed?

File: topcoffea/modules/corrections.py [hunk 4]
```diff
@@ -58,6 +59,7 @@ def get_method1a_wgt_doublewp(effA, effB, sfA, sfB, cutA, cutB, cutC):
 def btag_sf_eval(jet_collection,wp,year,method,syst):
     # Get the right sf json for the given year
 
+    dt_era = "Run3" if year[2] == "2" else "Run2"
```
Comment (kmohrman) [inline topcoffea/modules/corrections.py:62]: I would suggest replacing the strings with a bool `is_run3` to simplify statements later on.  That way we don't need to check for string agreement, and we only need one variable (since we can just do e.g., `if is_run3:` or whatever).

File: topcoffea/modules/createJSON.py [hunk 5]
```diff
@@ -85,7 +85,9 @@ def main():
         dicFiles = GetDatasetFromDAS(dataset, nFiles, options='file', withRedirector=prefix)
         files = [f[len(prefix):] for f in dicFiles['files']]
         files_with_prefix = dicFiles['files']
-
+        if 'root://cms-xrd-global.cern.ch//store/mc/Run3Summer22NanoAODv12/ZZZ_TuneCP5_13p6TeV_amcatnlo-pythia8/NANOAODSIM/130X_mcRun3_2022_realistic_v5-v2/50000/7c4f3eb2-3c7e-4c21-98ed-c1892bb3a057.root' in files_with_prefix:
+            print("ZZZ sample")
+            files_with_prefix.remove('root://cms-xrd-global.cern.ch//store/mc/Run3Summer22NanoAODv12/ZZZ_TuneCP5_13p6TeV_amcatnlo-pythia8/NANOAODSIM/130X_mcRun3_2022_realistic_v5-v2/50000/7c4f3eb2-3c7e-4c21-98ed-c1892bb3a057.root')
```
Comment (bryates) [inline topcoffea/modules/createJSON.py:90]: Are these lines for a bad file?

File: topcoffea/modules/CorrectedJetsFactory.py [hunk 6]
```diff
@@ -191,6 +189,11 @@ def __init__(self, name_map, jec_stack):
             self.jer_names = [name for name in self.jec_stack if isinstance(name, str) and ("Resolution" in name or "SF" in name)]
             self.junc_names = [name for name in self.jec_stack if isinstance(name, str) and ("Uncertainty" in name)]
             self.jec_names = [name for name in self.jec_stack if (name not in self.jer_names and name not in self.junc_names)]
+            #print("\n\n\n\n\n\n\n")
+            #print(self.jer_names)
+            #print(self.junc_names)
+            #print(self.jec_names)
+            #print("\n\n\n\n\n\n\n")
```
Comment (bryates) [inline topcoffea/modules/CorrectedJetsFactory.py:196]: Please remove

File: topcoffea/modules/CorrectedJetsFactory.py [hunk 7]
```diff
@@ -333,330 +336,328 @@ def build(self, jets, lazy_cache, isdata = False):
         out_dict[self.name_map["JetPt"] + "_jec"] = out_dict[self.name_map["JetPt"]]
         out_dict[self.name_map["JetMass"] + "_jec"] = out_dict[self.name_map["JetMass"]]
 
-        if not isdata: # Data does not need JER or uncertainties
```
Comment (bryates) [inline topcoffea/modules/CorrectedJetsFactory.py:336]: So we don't need a check for data here?

File: topcoffea/modules/createJSON.py [hunk 8]
```diff
@@ -128,7 +130,12 @@ def main():
             if not is_data:
                 if n_sum_of_lhe_weights is None:
                     n_sum_of_lhe_weights = list(i_sum_of_lhe_weights)
+                    print("\n n_sum_of_lhe_weights_new %s" % n_sum_of_lhe_weights)
```
Comment (bryates) [inline topcoffea/modules/createJSON.py:133]: Remove

File: topcoffea/modules/createJSON.py [hunk 9]
```diff
@@ -128,7 +130,12 @@ def main():
...
                 else:
+                    print('\n n_sum_of_lhe_weight: %s' % (n_sum_of_lhe_weights))
+                    print('\n i_sum_of_lhe_weight: %s' % (i_sum_of_lhe_weights))
+                    print('\n length of n sum of lhe weights: %s' % (len(n_sum_of_lhe_weights)))
+                    print('\n lenght of i sum of lhe weights: %s' % (len(i_sum_of_lhe_weights)))
```
Comment (bryates) [inline topcoffea/modules/createJSON.py:138]: Remove

**Review by bryates — COMMENTED** [review]
Thanks @anpicci, I just have couple of comments.

File: topcoffea/modules/event_selection.py (hunk 1 above)
Comment (anpicci) [reply topcoffea/modules/event_selection.py:12]: That's true, removed it

File: topcoffea/modules/createJSON.py (hunk 3 above)
Comment (anpicci) [reply topcoffea/modules/createJSON.py]: It is the ZZZ file with a malformed LHE branch, so Amanda skipped it

File: topcoffea/modules/corrections.py (hunk 4 above)
Comment (anpicci) [reply topcoffea/modules/corrections.py:62]: Sounds good!

File: topcoffea/modules/createJSON.py (hunk 5 above)
Comment (anpicci) [reply topcoffea/modules/createJSON.py:90]: Yes, the one missing an LHE branch

File: topcoffea/modules/CorrectedJetsFactory.py (hunk 7 above)
Comment (anpicci) [reply topcoffea/modules/CorrectedJetsFactory.py:336]: If it is data, `ApplyJetCorrections` from topeft won't require JER and JEC uncertainties to be computed. This is to expose to the user this choice, and don't hardcode it in an auxiliary class
https://github.com/TopEFT/topeft/blob/clib_run3/topeft/modules/corrections.py#L1427
https://github.com/TopEFT/topeft/blob/clib_run3/topeft/modules/corrections.py#L262-L285

File: topcoffea/modules/createJSON.py (hunk 8 above)
Comment (anpicci) [reply topcoffea/modules/createJSON.py:133]: done

**Comment by anpicci** [comment]
@bryates @kmohrman thank you for your feedback. I have addressed them, while merging the work done in [this branch](https://github.com/TopEFT/topcoffea/tree/coffeaJME) here. Originally, that branch was supposed to introduce upgrades for the version of `CorrectedJetsFactory` and `JECStack` ongoing into `Coffea` library. Given that I was able to deliver them sooner than expected, as agreed on MM, I proceeded to push these change to our repo as well

File: topcoffea/modules/corrections.py [hunk 10]
```diff
@@ -58,8 +58,9 @@ def get_method1a_wgt_doublewp(effA, effB, sfA, sfB, cutA, cutB, cutC):
 # Evaluate btag sf from central correctionlib json
 def btag_sf_eval(jet_collection,wp,year,method,syst):
     # Get the right sf json for the given year
+    is_run3 = year[2] == "2"
```
Comment (bryates) [inline topcoffea/modules/corrections.py:61]: Let's use a more clear check, like:
```suggestion
    is_run3 = yearstartswith("202")
```

File: topcoffea/params/xsec.yml (hunk 2 above)
Comment (kmohrman) [reply topcoffea/params/xsec.yml:212]: Looks like this comment is still open. Do you have references for these?

File: topcoffea/modules/createJSON.py (hunk 5 above)
Comment (kmohrman) [reply topcoffea/modules/createJSON.py:90]: Could you explain the handling here? So you are skipping a file? I am not sure we want to hard code this like this. This module is used by multiple analyses, and it's supposed to be fairly general. So this relatively silent skipping of a very particular root file is maybe not exactly how we want to handle it. 
Maybe would be better to pass an option to skip a file, or to just e.g. make this change locally if you don't want the file.

Also, what is wrong with the file?

File: topcoffea/modules/corrections.py (hunk 10 above)
Comment (kmohrman) [reply topcoffea/modules/corrections.py]: I would actually suggest checking for exact string agreement. E.g. `if year in ["2022","2022EE",...]`. I think it's safer

File: topcoffea/params/xsec.yml (hunk 2 above)
Comment (anpicci) [reply topcoffea/params/xsec.yml:212]: Done with the forthcoming commit

File: topcoffea/modules/createJSON.py (hunk 5 above)
Comment (anpicci) [reply topcoffea/modules/createJSON.py:90]: According to GEN contacts, this is the issue. Given that, we can remove these lines, not using the ZZZ sample for now.
What do you think @bryates @kmohrman ?

Screenshot 2024-11-05 at 10 05 42

File: topcoffea/modules/corrections.py (hunk 10 above)
Comment (anpicci) [reply topcoffea/modules/corrections.py]: I am in favor of the `.startswith` solution. Given that Run2 years will always return `is_run3 = False` and future Run3 samples will automatically make this boolean `True`, I think that solution will work in the near future, without the need to re-modify this line (as instead, it would happen with the list solution). I don't see any particular safety issues that the list implementation can prevent and the `.startswith` solution cannot

File: topcoffea/modules/event_selection.py [hunk 11]
```diff
@@ -9,7 +9,6 @@
 def passes_trg_inlst(events,trg_name_lst):
     tpass = np.zeros_like(np.array(events.MET.pt), dtype=bool)
     trg_info_dict = events.HLT
-
```
[the remark this answers is not in the document]
Comment (anpicci) [reply topcoffea/modules/event_selection.py]: Done with the forthcoming commit

File: topcoffea/modules/createJSON.py (hunk 5 above)
Comment (bryates) [reply topcoffea/modules/createJSON.py:90]: I guess this is fine for a temporary solution. The email from the GEN contacts makes it sound like it _might_ be fixable, so I opened an issue to remind us to check on this later.

**Review by bryates — APPROVED** [review]
I think I'm happy with these changes now. Once @kmohrman signs off, we should be good to merge.

**Comment by kmohrman** [comment]
> I think I'm happy with these changes now. Once @kmohrman signs off, we should be good to merge.

Thanks. Since this touches a lot of Run 3 functions, we will need to run this through `ewkcoffea` and validate that there are not unexpected changes.

Things are especially busy right now, so we probably won't get to this till next week. @bryates is that ok, or would that block progress on your end?

**Comment by bryates** [comment]
> > I think I'm happy with these changes now. Once @kmohrman signs off, we should be good to merge.
> 
> Thanks. Since this touches a lot of Run 3 functions, we will need to run this through `ewkcoffea` and validate that there are not unexpected changes.
> 
> Things are especially busy right now with the review of our analysis, so we probably won't get to this till next week. @bryates is that ok, or would that block progress on your end?

That's fine with me.

**Comment by anpicci** [comment]
@bryates should we notify Kevin about this?

**Comment by bryates** [comment]
> @bryates should we notify Kevin about this?

I was going to make a quick fix on TopEFT to check out this branch. That will fix it CI.

**Comment by kmohrman** [comment]
> TopEFT/topeft#436

Ah, ok. So the issue is not particularly related to this PR. Thanks for explaining.

**Comment by bryates** [comment]
Just checking, @kmohrman and @MatthewDittrich have you had time to test this PR in your repo?

**Comment by bryates** [comment]
> > Just checking, @kmohrman and @MatthewDittrich have you had time to test this PR in your repo?
> 
> We have not yet. Is it blocking progress on your end? If so, I will prioritize testing it before Monday.
> 
> However, if it's not blocking anything right now, then I might request pushing back the review by another week. We're in the middle of the ARC review for the WWZ analysis, so things are a bit busy. And it's a difficult time to be pushing large changes to the `topcoffea` backend. We will need to be _super_ sure it's not changing the result at all in any unexpected ways, so we'll need to give it a careful review.
> 
> So, please just let me know, and I will prioritize accordingly.

No worries. This isn't blocking any progress on our end, so take your time 😄

**Comment by bryates** [comment]
> > > Just checking, @kmohrman and @MatthewDittrich have you had time to test this PR in your repo?
> > 
> > 
> > We have not yet. Is it blocking progress on your end? If so, I will prioritize testing it before Monday.
> > However, if it's not blocking anything right now, then I might request pushing back the review by another week. We're in the middle of the ARC review for the WWZ analysis, so things are a bit busy. And it's a difficult time to be pushing large changes to the `topcoffea` backend. We will need to be _super_ sure it's not changing the result at all in any unexpected ways, so we'll need to give it a careful review.
> > So, please just let me know, and I will prioritize accordingly.
> 
> No worries. This isn't blocking any progress on our end, so take your time 😄

@kmohrman @anpicci @MatthewDittrich any updates on this PR?

**Comment by kmohrman** [comment]
> > > We have not yet. Is it blocking progress on your end? If so, I will prioritize testing it before Monday.
> > > However, if it's not blocking anything right now, then I might request pushing back the review by another week. We're in the middle of the ARC review for the WWZ analysis, so things are a bit busy. And it's a difficult time to be pushing large changes to the `topcoffea` backend. We will need to be _super_ sure it's not changing the result at all in any unexpected ways, so we'll need to give it a careful review.
> > > So, please just let me know, and I will prioritize accordingly.
> > 
> > 
> > No worries. This isn't blocking any progress on our end, so take your time 😄
> 
> @kmohrman @anpicci @MatthewDittrich any updates on this PR?

Hello @bryates. Sorry I did not realize that the `topeft` master depends on this branch of `topcoffea`, and that this is causing issues for the CCL's CI. Ideally before merging, the changes would need to be fully validated in `ewkcoffea`, especially since the WWZ analysis is currently going through preapproval/unblidning. The fact that we happen to be in this particularly crucial time of the analysis also makes it difficult to find the time to test a large PR. 

Because of the situation, @klannon suggested that we go ahead and merge this PR now, and that for `ewkcoffea`, we can temporarily move to [a branch of `topcoffea`](https://github.com/TopEFT/topcoffea/tree/pin_dec06_2024) (which corresponds to the current master). Once the crunch time is over for WWZ, we will do a full validation of the changes in this PR (especially the Run 3 stuff). This is not necessarily ideal, but seems to be the best path forward given the situation.  

**TL;DR**: I'm happy for this PR to be merged now (and `ewkcoffea` is already moved to the `pin_dec06_2024` branch). But upon its merge, I will open an issue to remind us about the pending validation.

**Comment by bryates** [comment]
> > > > We have not yet. Is it blocking progress on your end? If so, I will prioritize testing it before Monday.
> > > > However, if it's not blocking anything right now, then I might request pushing back the review by another week. We're in the middle of the ARC review for the WWZ analysis, so things are a bit busy. And it's a difficult time to be pushing large changes to the `topcoffea` backend. We will need to be _super_ sure it's not changing the result at all in any unexpected ways, so we'll need to give it a careful review.
> > > > So, please just let me know, and I will prioritize accordingly.
> > > 
> > > 
> > > No worries. This isn't blocking any progress on our end, so take your time 😄
> > 
> > @kmohrman @anpicci @MatthewDittrich any updates on this PR?
> 
> Hello @bryates. Sorry I did not realize that the `topeft` master depends on this branch of `topcoffea`, and that this is causing issues for the CCL's CI. Ideally before merging, the changes would need to be fully validated in `ewkcoffea`, especially since the WWZ analysis is currently going through preapproval/unblidning. The fact that we happen to be in this particularly crucial time of the analysis also makes it difficult to find the time to test a large PR. 
> 
> Because of the situation, @klannon suggested that we go ahead and merge this PR now, and that for `ewkcoffea`, we can temporarily move to [a branch of `topcoffea`](https://github.com/TopEFT/topcoffea/tree/pin_dec06_2024) (which corresponds to the current master). Once the crunch time is over for WWZ, we will do a full validation of the changes in this PR (especially the Run 3 stuff). This is not necessarily ideal, but seems to be the best path forward given the situation.  
> 
> **TL;DR**: I'm happy for this PR to be merged now (and `ewkcoffea` is already moved to the `pin_dec06_2024` branch). But upon its merge, I will open an issue to remind us about the pending validation. 

Sounds good. I'll merge it now.
